File: libgomp/testsuite/libgomp.c/examples-4/simd-1.c
/* { dg-do run } */
/* { dg-additional-options "-msse2" { target sse2_runtime } } */
/* { dg-additional-options "-mavx" { target avx_runtime } } */

#define N 100
#define OFF 32
#define EPS 0.0000000000000001

#include <stdlib.h>

void init(double *a, double *a_ref, double *b, double *c, int n, int ioff)
{
  int i;
  for ( i = 0; i < n; i++ )
  {
    a[i] = i*i;
    a_ref[i] = a[i];
    b[i] = i+i;
  }

  int s = -1;
  for ( i = 0; i < n+ioff; i++ )
  {
    c[i] = s*3;
    s = -s;
  }
}

void star( double *a, double *b, double *c, int n, int *ioff )
{
  int i;
  #pragma omp simd
  for ( i = 0; i < n; i++ )
    a[i] *= b[i] * c[i+ *ioff];
}

void star_ref( double *a, double *b, double *c, int n, int *ioff )
{
  int i;
  for ( i = 0; i < n; i++ )
    a[i] *= b[i] * c[i+ *ioff];
}

void check (double *a, double *b)
{
  int i;
  for (i = 0; i < N; i++)
    if (a[i] - b[i] > EPS || b[i] - a[i] > EPS)
      abort ();
}

int main ()
{
  double a[N], a_ref[N], b[N], c[N+OFF];
  int ioff = OFF;

  init(a, a_ref, b, c, N, ioff);

  star(a, b, c, N, &ioff);
  star_ref(a_ref, b, c, N, &ioff);

  check(a, a_ref);

  return 0;
}

File: libgomp/testsuite/libgomp.fortran/examples-4/simd-1.f90
! { dg-do run }
! { dg-additional-options "-msse2" { target sse2_runtime } }
! { dg-additional-options "-mavx" { target avx_runtime } }

module SIMD1_mod
contains
  subroutine init (a, a_ref, b, c, n, ioff_ptr)
    double precision :: a(*), a_ref(*), b(*), c(*)
    integer          :: n, i, s
    integer, pointer :: ioff_ptr

    s = -1
    do i = 1, n
      a(i) = i * i * s
      a_ref(i) = a(i)
      b(i) = i + i
    end do

    do i = 1, n+ioff_ptr
      c(i) = i * 3
    end do

  end subroutine

  subroutine check (a, b, n)
    integer :: i, n
    double precision, parameter :: EPS = 0.0000000000001
    double precision :: diff, a(*), b(*)
    do i = 1, n
      diff = a(i) - b(i)
      if (diff > EPS .or. -diff > EPS) call abort
    end do
  end subroutine

  subroutine star(a, a_ref, b, c, n, ioff_ptr)
     double precision :: a(*), a_ref(*), b(*), c(*)
     integer          :: n, i
     integer, pointer :: ioff_ptr

     call init (a, a_ref, b, c, n, ioff_ptr)

     !$omp simd
     do i = 1,n
        a(i) = a(i) * b(i) * c(i+ioff_ptr)
     end do

     do i = 1,n
        a_ref(i) = a_ref(i) * b(i) * c(i+ioff_ptr)
     end do

     call check (a, a_ref, n)

  end subroutine
end module

program SIMD1
  use SIMD1_mod, only : star
  double precision :: a(128), a_ref(128), b(128), c(144)
  integer, pointer:: ioff_ptr
  integer, target:: offset

  offset = 16
  ioff_ptr => offset

  call star (a, a_ref, b, c, 128, ioff_ptr)

end program
